magma/passes/insert_coreir_wires.py

```diff
@@ -0,0 +1,103 @@
+from .passes import DefinitionPass
+from ..digital import Digital
+from ..generator import Generator
+from ..circuit import Circuit, DeclareCoreirCircuit, _PlacerBlock
+from ..t import In, Out
+from ..digital import Digital
+from ..array import Array
+from ..tuple import Tuple
+from ..bits import Bits
+from ..ref import NamedRef
+
+
+def simulate_wire(self, value_store, state_store):
+    value_store.set_value(self.O, value_store.get_value(self.I))
+
+
+class Wire(Generator):
+    @staticmethod
+    def generate(T):
+        if issubclass(T, Digital):
+            coreir_lib = "corebit"
+            coreir_genargs = None
+        else:
+            coreir_lib = "coreir"
+            width = T.flat_length()
+            T = Bits[width]
+            coreir_genargs = {"width": width}
+        return DeclareCoreirCircuit(
+            "Wire", 'I', In(T), 'O', Out(T),
+            simulate=simulate_wire,
+            coreir_name="wire",
+            coreir_lib=coreir_lib,
+            coreir_genargs=coreir_genargs
+        )
+
+
+def _sanitize_name(name):
+    return name.replace("[", "_").replace("]", "")
+
+
+class InsertCoreIRWires(DefinitionPass):
+    def __init__(self, main):
+        super().__init__(main)
+        self.seen = set()
+        self.wire_map = {}
+
+    def insert_wire(self, value, definition):
+        if value.is_mixed():
+            # mixed children
+            for child in value:
+                if not child.is_output():
+                    self.insert_wire(child, definition)
+            return
+        if value in self.seen:
+            # In the case of inouts, we may see more than once
+            return
+        self.seen.add(value)
+        driver = value.value()
+        while driver is not None and driver.name.anon():
+            driver = driver.value()
+        if driver is None or driver.is_output():
+            return
+
+        value.unwire(driver)
+        T = type(driver)
+        if driver not in self.wire_map:
+            driver_name = driver.name.qualifiedname("_")
+            value_name = value.name.qualifiedname("_")
+            driver_name = _sanitize_name(driver_name)
+            value_name = _sanitize_name(value_name)
+
+            # Rename driver so it doesn't conflict with new wire instance name
+            if isinstance(driver.name, NamedRef):
+                driver.name.name = "_" + driver.name.name
+            name = f"{driver_name}"
+            with _PlacerBlock(definition._placer):
```

Comment:
this is better but I think having an explicit open() function in circuit is good as an external API, and keeping _PlacerBlock as an internal to the circuit implementation. It will pretty much dispatch to this but hides those details. Also, I think all the wiring should happen in that with context. Right now it doesn't make a difference but I could forsee checking that wiring always happens inside in the right context, and either way it is good practice.